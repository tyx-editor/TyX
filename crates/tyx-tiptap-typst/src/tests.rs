use std::path::PathBuf;

use ecow::eco_format;
use serde::Serialize;
use tinymist_project::LspWorld;
use typst::{WorldExt, syntax::Span};

use crate::{SyntaxTree, syntax::SyntaxDecl};

/// Makes snapshot tests for the given function.
/// The fixtures used for snapshot testing are in the `src/fixtures/{name}`
/// directory.
fn snapshot_testing(name: &str, f: &impl Fn(LspWorld, PathBuf)) {
    tinymist_tests::snapshot_testing!(name, |verse, path| {
        f(verse.snapshot(), path);
    });
}

#[test]
fn convert() {
    snapshot_testing("integration", &|world, _path| {
        insta::assert_snapshot!(conv(world));
    });
}

/// Converts the given main document in the world to a string.
fn conv(world: LspWorld) -> String {
    let doc = crate::convert(world);
    let node = doc.map(|doc| doc.content);
    serialize(&node)
}

#[test]
fn instrument() {
    snapshot_testing("instrument", &|world, _path| {
        insta::assert_snapshot!(instr(world));
    });
}

/// Converts the given main document in the world to a string.
fn instr(world: LspWorld) -> String {
    let doc = crate::instrument(&world).map(|(doc, _)| resolve_span(&world, doc));
    serialize(&doc)
}

fn resolve_span(world: &LspWorld, doc: SyntaxTree) -> SyntaxTree {
    match doc {
        SyntaxTree::Markup(decl) => {
            let span = u64::from_str_radix(&decl.span, 16).unwrap();
            let range = world
                .range(Span::from_raw(span.try_into().unwrap()))
                .unwrap();
            SyntaxTree::Markup(SyntaxDecl {
                span: eco_format!("{range:?}"),
                content: decl.content,
                semi: decl.semi,
            })
        }
        SyntaxTree::Decl(decl) => {
            let span = u64::from_str_radix(&decl.span, 16).unwrap();
            let range = world
                .range(Span::from_raw(span.try_into().unwrap()))
                .unwrap();
            SyntaxTree::Decl(SyntaxDecl {
                span: eco_format!("{range:?}"),
                content: decl.content,
                semi: decl.semi,
            })
        }
        SyntaxTree::Content(children) => {
            let children = children
                .into_iter()
                .map(|child| resolve_span(world, child))
                .collect();
            SyntaxTree::Content(children)
        }
    }
}

// to pretty with indent 1
fn serialize<T: Serialize>(value: &T) -> String {
    let mut w = Vec::new();
    let mut s = serde_json::Serializer::with_formatter(
        &mut w,
        serde_json::ser::PrettyFormatter::with_indent(b" "),
    );
    value.serialize(&mut s).expect("failed to serialize");
    String::from_utf8(w).expect("invalid utf8")
}
