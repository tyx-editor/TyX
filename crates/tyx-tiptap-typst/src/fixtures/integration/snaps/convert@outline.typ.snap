---
source: crates/tyx-tiptap-typst/src/tests.rs
expression: conv(world)
input_file: crates/tyx-tiptap-typst/src/fixtures/integration/outline.typ
---
{
 "type": "doc",
 "content": [
  {
   "type": "heading",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Contents"
      }
     ]
    }
   ],
   "dir": null,
   "level": 2
  },
  {
   "type": "heading",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Heading 1"
      }
     ]
    }
   ],
   "dir": null,
   "level": 3
  },
  {
   "type": "heading",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Heading 2"
      }
     ]
    }
   ],
   "dir": null,
   "level": 4
  },
  {
   "type": "heading",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Heading 1"
      }
     ]
    }
   ],
   "dir": null,
   "level": 3
  },
  {
   "type": "heading",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Heading 2"
      }
     ]
    }
   ],
   "dir": null,
   "level": 4
  },
  {
   "type": "paragraph",
   "content": [
    {
     "type": "mark",
     "type": "link",
     "content": [
      {
       "type": "text",
       "content": [
        {
         "type": "plain",
         "content": "This is a link to example.com"
        }
       ]
      }
     ],
     "class": null,
     "href": "https://example.com"
    }
   ],
   "dir": null
  },
  {
   "type": "paragraph",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Inline"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": " "
      }
     ]
    },
    {
     "type": "mark",
     "type": "code",
     "content": [
      {
       "type": "plain",
       "content": "code"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": " "
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "has"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": " "
      }
     ]
    },
    {
     "type": "mark",
     "type": "code",
     "content": [
      {
       "type": "plain",
       "content": "back-ticks around"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": " "
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "it."
      }
     ]
    }
   ],
   "dir": null
  },
  {
   "type": "code_block",
   "content": [
    {
     "type": "plain",
     "content": "using System.IO.Compression;\n\n#pragma warning disable 414, 3021\n\nnamespace MyApplication\n{\n    [Obsolete(\"...\")]\n    class Program : IInterface\n    {\n        public static List<int> JustDoIt(int count)\n        {\n            Console.WriteLine($\"Hello {Name}!\");\n            return new List<int>(new int[] { 1, 2, 3 })\n        }\n    }\n}"
    }
   ],
   "language": "cs"
  },
  {
   "type": "paragraph",
   "content": [
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "Math inline"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": ":"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": "and block"
      }
     ]
    },
    {
     "type": "text",
     "content": [
      {
       "type": "plain",
       "content": ":"
      }
     ]
    }
   ],
   "dir": null
  },
  {
   "type": "paragraph",
   "content": [],
   "dir": null
  },
  {
   "type": "bullet_list",
   "content": [
    {
     "type": "list_item",
     "content": [
      {
       "type": "paragraph",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "First item"
          }
         ]
        }
       ],
       "dir": null
      }
     ]
    },
    {
     "type": "list_item",
     "content": [
      {
       "type": "paragraph",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "Second item"
          }
         ]
        }
       ],
       "dir": null
      },
      {
       "type": "ordered_list",
       "content": [
        {
         "type": "list_item",
         "content": [
          {
           "type": "paragraph",
           "content": [
            {
             "type": "text",
             "content": [
              {
               "type": "plain",
               "content": "First sub-item"
              }
             ]
            }
           ],
           "dir": null
          }
         ]
        },
        {
         "type": "list_item",
         "content": [
          {
           "type": "paragraph",
           "content": [
            {
             "type": "text",
             "content": [
              {
               "type": "plain",
               "content": "Second sub-item"
              }
             ]
            }
           ],
           "dir": null
          },
          {
           "type": "bullet_list",
           "content": [
            {
             "type": "list_item",
             "content": [
              {
               "type": "paragraph",
               "content": [
                {
                 "type": "text",
                 "content": [
                  {
                   "type": "plain",
                   "content": "First sub-sub-item"
                  }
                 ]
                }
               ],
               "dir": null
              }
             ]
            }
           ]
          }
         ]
        }
       ],
       "start": 1,
       "type": null
      }
     ]
    }
   ]
  },
  {
   "type": "paragraph",
   "content": [],
   "dir": null
  },
  {
   "type": "table",
   "content": [
    {
     "type": "table_row",
     "content": [
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "0"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "1"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "2"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "3"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "4"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "5"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "6"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "7"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "8"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "9"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "10"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "11"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "12"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "13"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "14"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "15"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "16"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "17"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "18"
          }
         ]
        }
       ],
       "colwidth": null
      },
      {
       "type": "table_header",
       "content": [
        {
         "type": "text",
         "content": [
          {
           "type": "plain",
           "content": "19"
          }
         ]
        }
       ],
       "colwidth": null
      }
     ]
    }
   ]
  }
 ]
}
